api/pkg/auth/service/service.go

```diff
@@ -92,58 +92,159 @@ func createChecksum(token string) string {
-	err := q.First(&user).Error
+	// If email doesn't exists in users table
 	if err != nil {
+		// Check whether username and provider are matching
+		accountQuery := r.db.Model(&model.Account{}).Where("user_name = ?", gitUser.NickName).Where("provider = ?", provider)
+		err = accountQuery.First(&acc).Error
+
 		// If user doesn't exist, create a new record
 		if err == gorm.ErrRecordNotFound {
+			if user, err = r.insertIntoUsersTable(gitUser, code); err != nil {
+				r.log.Error(err)
+			}
 
-			user.GithubName = ghUser.Name
-			user.GithubLogin = ghUser.NickName
-			user.Type = model.NormalUserType
-			user.AvatarURL = ghUser.AvatarURL
-			user.Code = code
+			if err = r.insertIntoAccountsTable(gitUser, provider, user.ID); err != nil {
+				r.log.Error(err)
+				return err
+			}
+		} else {
+			// Account exists
+			// Update the user table with the email
+			if err := r.db.Model(&model.User{}).Where("id = ?", acc.UserID).
+				Updates(model.User{Code: code, Email: gitUser.Email, Type: model.NormalUserType}).Error; err != nil {
+				r.log.Error(err)
+				return err
+			}
 
-			err = r.db.Create(&user).Error
-			if err != nil {
+			// Update the AvatarUrl and Name in Accounts table
+			if err := updateAccountDetails(accountQuery, acc,
+				model.Account{AvatarURL: gitUser.AvatarURL, Name: gitUser.Name}); err != nil {
 				r.log.Error(err)
 				return err
 			}
 		}
-	} else {
-		if err := r.db.Model(&model.User{}).Where("github_login = ?", ghUser.NickName).Update("code", code).Error; err != nil {
+	} else { // when the email of user already exists
+		// Update the users table with the auth code
+		if err := userQuery.Update("code", code).Error; err != nil {
+			r.log.Error(err)
+			return err
+		}
+
+		// Check for the account on the basis of user_id and provider
+		accountQuery := r.db.Model(&model.Account{}).Where("user_id = ?", user.ID).Where("provider = ?", provider)
+		err = accountQuery.First(&acc).Error
+
+		// If not found then create a new entry in accounts table
+		if err == gorm.ErrRecordNotFound {
+			if err = r.insertIntoAccountsTable(gitUser, provider, user.ID); err != nil {
+				r.log.Error(err)
+				return err
+			}
+			return nil
+		} else if err != nil {
+			r.log.Error(err)
+			return err
+		}
+
+		// If account found then update the details of the user
+		if err := updateAccountDetails(accountQuery, acc,
+			model.Account{AvatarURL: gitUser.AvatarURL, Name: gitUser.Name, UserName: gitUser.NickName}); err != nil {
 			r.log.Error(err)
 			return err
 		}
+
 	}
 
-	// If user already exists i.e if user is added through config
-	// and is missing git username and avatarUrl, then update the code
-	user.Code = code
+	return nil
+}
 
-	// User already exist, check if GitHub Name is empty
-	// If Name is empty, then user is inserted through config.yaml
-	// Update user with remaining details
+func updateAccountDetails(accountQuery *gorm.DB, existingAccountDetails model.Account, newAccountDetails model.Account) error {
```

Comment:
Can we create a struct which has the params of this functions as fields of struct and whenever we call this function we create an object of this struct and call this function with a receiver object for example `func (obj structType) updateAccountDetails() {}`

Reply:
This is a function that is not used at multiple places and also three params are struct only, its fine to keep it as it is for now